### PR TITLE
Remove deregistered objects from the inactive overload cache

### PR DESCRIPTION
**Pull request by BetsyMcPhail**
Fixes issue reported in https://github.com/RobotLocomotion/drake/issues/11424

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)

### CODE REVIEW

**Review by EricCousineau-TRI — COMMENTED** [review]
+@ericcousineau-tri for review

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-:-LoLWac05_lS9sah-_ZC:b7dr7wo)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @EricCousineau-TRI)

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: all files reviewed, 3 unresolved discussions (waiting on @BetsyMcPhail)

___
*[include/pybind11/pybind11.h, line 2413 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_ix832_XDf5KPxVfI:-Lo_ix832_XDf5KPxVfJ:b-99dhct) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/pybind11.h#L2413)):*
> ```objc
>     if (!self)
>         return function();
>     handle type = self.get_type();
> ```

nit CI errors are happening due to this being unused. This error should disappear if you try the suggestion; if that doesn't work, then you'll want to remove this.
___
*[include/pybind11/pybind11.h, line 2414 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_icCO0eRVwzEwqke7:-Lo_icCPDURiznBu_poL:bvn8354) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/pybind11.h#L2414)):*
> ```objc
>         return function();
>     handle type = self.get_type();
>     auto key = std::make_pair(self.ptr(), name);
> ```

Rather than use the object pointer and require removal upon object destruction, can you try my suggestion from this comment?
https://github.com/RobotLocomotion/drake/issues/11424#issuecomment-527992628
___
*[include/pybind11/detail/class.h, line 228 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_ipEDCYTs6GlomGkx:-Lo_ipEDCYTs6GlomGky:bmpnw0j) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/detail/class.h#L228)):*
> ```objc
>         if (self == it->second && Py_TYPE(self) == Py_TYPE(it->second)) {
>             registered_instances.erase(it);
>             PyObject *to_remove = (PyObject *)it->second;
> ```

When you try my above suggestion, can you remove this code?

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: all files reviewed, 5 unresolved discussions (waiting on @BetsyMcPhail)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jHkABFdhd1SkNDy3:-Lo_j42UETLdqYSUjuFx:bive7ky) (no related file):*
I think this could be testable with the following procedure. Per the root-cause analysis in the Drake issue:
* Have two functions that creates a type that has a pybind "vtable", and return the type
* Call one function, and call it's overload.
* Delete the type reference, garbage collect, ensure that it was deleted, then create the other type, and check.

---
*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jn4T3g1jU3NRzQyg:-Lo_jn4U3ETKJtpXIrYv:bwumv8o) (no related file):*
I'm going to briefly try out my own suggestions; will post back to this discussion item.

---

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[include/pybind11/pybind11.h, line 2413 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_ix832_XDf5KPxVfI:-Lo_m9b09uB2p6b_CuT3:b-rzteoe) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/pybind11.h#L2413)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit CI errors are happening due to this being unused. This error should disappear if you try the suggestion; if that doesn't work, then you'll want to remove this.
</blockquote></details>

Done. (Fixed in separate push.)
___
*[include/pybind11/pybind11.h, line 2414 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_icCO0eRVwzEwqke7:-Lo_mBSb8RaCpCksMbVy:b-3cwt4n) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/pybind11.h#L2414)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Rather than use the object pointer and require removal upon object destruction, can you try my suggestion from this comment?
https://github.com/RobotLocomotion/drake/issues/11424#issuecomment-527992628
</blockquote></details>

This does look like it'd be rather complicated, and would require non-trivial changes to the internals - or it would require on-purpose memory leaks to concatenate the full type name into the `key`.

However, the current workaround does seem a bit expensive, and I'm not sure it would get accepted in upstream as-is; perhaps it would get accepted with the internals modification, or simply concatenating.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jn4T3g1jU3NRzQyg:-Lob7jb1Byih5OFHHDJa:b-fs8r4t) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I'm going to briefly try out my own suggestions; will post back to this discussion item.
</blockquote></details>

Tried making unittest, but encountered a different error; will post about that separately, possibly directly in upstream.

Here's the failure reproduced directly in upstream:
https://github.com/EricCousineau-TRI/pybind11/commit/97260868f5fb0f1a0ba81c38b173bee99b6c71c4

I'll apply the patch here, and push it.

---

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 0 of 2 files reviewed, 4 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jn4T3g1jU3NRzQyg:-LobKve02VijJzuErqo8:bv0s6uo) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Tried making unittest, but encountered a different error; will post about that separately, possibly directly in upstream.

Here's the failure reproduced directly in upstream:
https://github.com/EricCousineau-TRI/pybind11/commit/97260868f5fb0f1a0ba81c38b173bee99b6c71c4

I'll apply the patch here, and push it.
</blockquote></details>

(To clarify, I've created both a test which exercises this failure, as well as finding another error...)

---

**Review by EricCousineau-TRI — APPROVED** [review]
:lgtm: (pending CI)

> Reviewed 2 of 2 files at r2, 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-:-LobTIv0BIpRWGearTn8:b-lj32db)* status: :shipit: complete! all files reviewed, all discussions resolved

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jHkABFdhd1SkNDy3:-LobT43d7XPldJsYUl3G:b-na16l3) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I think this could be testable with the following procedure. Per the root-cause analysis in the Drake issue:
* Have two functions that creates a type that has a pybind "vtable", and return the type
* Call one function, and call it's overload.
* Delete the type reference, garbage collect, ensure that it was deleted, then create the other type, and check.
</blockquote></details>

OK Retracting. We can see what pybind11 authors might suggest if/when they have the bandwidth. Added TODOs.

---
*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_jn4T3g1jU3NRzQyg:-LobTAJN3VkSAlSNyW1Y:b-896fix) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

(To clarify, I've created both a test which exercises this failure, as well as finding another error...)
</blockquote></details>

Done.

---
___
*[include/pybind11/pybind11.h, line 2414 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_icCO0eRVwzEwqke7:-LobTBZB7JBvyz3tdmqg:bmu4mm1) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/pybind11.h#L2414)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

This does look like it'd be rather complicated, and would require non-trivial changes to the internals - or it would require on-purpose memory leaks to concatenate the full type name into the `key`.

However, the current workaround does seem a bit expensive, and I'm not sure it would get accepted in upstream as-is; perhaps it would get accepted with the internals modification, or simply concatenating.
</blockquote></details>

OK Retracting given that there's now a TODO.
___
*[include/pybind11/detail/class.h, line 228 at r1](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lo_ipEDCYTs6GlomGkx:-LobTEckDuA5u5UTfwxU:b-wxo176) ([raw file](https://github.com/robotlocomotion/pybind11/blob/5661566fd63ed6b3ba56ddacb4380d1448df0f92/include/pybind11/detail/class.h#L228)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

When you try my above suggestion, can you remove this code?
</blockquote></details>

OK Retracting.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[tests/test_class.py, line 312 at r4](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp3nt3O4dHn2_mULwUE:-Lp3nt3O4dHn2_mULwUF:b-5wdgr3) ([raw file](https://github.com/robotlocomotion/pybind11/blob/351c186ce4aef7418f998c640dd6cb95f3b14504/tests/test_class.py#L312)):*
> ```Python
>     del Child1
>     pytest.gc_collect()
>     assert wref() is None
> ```

Sorry for not fixing this earlier; I think this will still ultimately fail CI.

The right fix would be to possibly disable this test unless it's under CPython, Python 3. (I think `pytest` has some `skipIf` mechanism can you can combine with info from `sys`.)

Let me know if you'd like me to make that change.

**Review by BetsyMcPhail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @EricCousineau-TRI)

___
*[tests/test_class.py, line 312 at r4](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp3nt3O4dHn2_mULwUE:-Lp3qS6x7lpBDZs2Yadz:b90zd55) ([raw file](https://github.com/robotlocomotion/pybind11/blob/351c186ce4aef7418f998c640dd6cb95f3b14504/tests/test_class.py#L312)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Sorry for not fixing this earlier; I think this will still ultimately fail CI.

The right fix would be to possibly disable this test unless it's under CPython, Python 3. (I think `pytest` has some `skipIf` mechanism can you can combine with info from `sys`.)

Let me know if you'd like me to make that change.
</blockquote></details>

No problem. I figured I'd make the quick fix first and then dig into any remaining CI failures.  I can make the other change since I'm already working in this file. I'll let you know if I run into any issues.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[tests/test_class.py, line 312 at r4](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp3nt3O4dHn2_mULwUE:-Lp45WlsE6TmKTVFNBp5:b-32rklt) ([raw file](https://github.com/robotlocomotion/pybind11/blob/351c186ce4aef7418f998c640dd6cb95f3b14504/tests/test_class.py#L312)):*
<details><summary><i>Previously, BetsyMcPhail (Betsy McPhail) wrote…</i></summary><blockquote>

No problem. I figured I'd make the quick fix first and then dig into any remaining CI failures.  I can make the other change since I'm already working in this file. I'll let you know if I run into any issues.
</blockquote></details>

Sounds good. From looking at the prior build: https://travis-ci.com/RobotLocomotion/pybind11/builds/128189983

Looks like this test should only run if it's on Linux, CPython, and not a debug build.
Also, it looks like c++11 / gcc4.8 also fails in Python 2, so perhaps constrain it only run for Python 3?

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[tests/test_class.py, line 296 at r5](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp4J34HDIctea5Z-zUS:-Lp4J34ICvhSD9vo0SKf:b9uc6ta) ([raw file](https://github.com/robotlocomotion/pybind11/blob/356f862dea3ced15149d9d6f38de84f37f7de956/tests/test_class.py#L296)):*
> ```Python
> @pytest.unsupported_on_pypy
> @pytest.mark.skipif(sys.platform != "linux", reason="only runs on Linux")
> @pytest.mark.skipif(sys.version_info < (3, 0), reason="requires python3")
> ```

nit You may to also check if this is a debug build.
From looking, perhaps the check is `if "d" in sys.abiflags`

This may be valid only in Python 3 (3.2 per the PEP), so perhaps add it to the existing stuff:

```
skipif(sys.version_info < (3, 0) or "d" in sys.abiflags, reason="requires python3 in release mode")
```

PEP: https://www.python.org/dev/peps/pep-3149/

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[tests/test_class.py, line 296 at r6](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp53GIX3Nhi5aYbgD5m:-Lp53GIX3Nhi5aYbgD5n:b-my5qxa) ([raw file](https://github.com/robotlocomotion/pybind11/blob/cd987162734467d8635cb324154a8b8d223375ee/tests/test_class.py#L296)):*
> ```Python
> 
> 
> @pytest.unsupported_on_pypy
> ```

:( It looks like this is very platform dependent; this failure seems like it's hard to check in a way that's not CI-specific (e.g. checking for environment variables):
https://travis-ci.com/RobotLocomotion/pybind11/jobs/236907166 (fails, but is CPython, Python 3, non-debug, Linux)

Perhaps this should just be skipped?
```
pytest.mark.skip(reason="Generally reproducible in CPython, Python 3, non-debug, on Linux. However, it is hard to pin this down for CI.")
```

**Review by BetsyMcPhail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @EricCousineau-TRI)

___
*[tests/test_class.py, line 296 at r6](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp53GIX3Nhi5aYbgD5m:-Lp8spLrBTY0hCEEIGZB:b-dn5qng) ([raw file](https://github.com/robotlocomotion/pybind11/blob/cd987162734467d8635cb324154a8b8d223375ee/tests/test_class.py#L296)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

:( It looks like this is very platform dependent; this failure seems like it's hard to check in a way that's not CI-specific (e.g. checking for environment variables):
https://travis-ci.com/RobotLocomotion/pybind11/jobs/236907166 (fails, but is CPython, Python 3, non-debug, Linux)

Perhaps this should just be skipped?
```
pytest.mark.skip(reason="Generally reproducible in CPython, Python 3, non-debug, on Linux. However, it is hard to pin this down for CI.")
```
</blockquote></details>

Unfortunately, I think you're right. It's going to be hard to define a combination that works reliably.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp8zZ707lWngEEXElV0:-Lp8zZ707lWngEEXElV1:bb49vtr) (no related file):*
I will file the Drake PR that tests using this commit.

---

**Comment by EricCousineau-TRI** [comment]
Almost there! One last lint error:

```
$ flake8
./tests/test_class.py:2:1: F401 'sys' imported but unused
import sys
```

**Review by BetsyMcPhail — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 2 unresolved discussions (waiting on @BetsyMcPhail and @EricCousineau-TRI)

___
*[tests/test_class.py, line 296 at r6](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp53GIX3Nhi5aYbgD5m:-LpYjIh09Csd0jG6RGdx:b-896fix) ([raw file](https://github.com/robotlocomotion/pybind11/blob/cd987162734467d8635cb324154a8b8d223375ee/tests/test_class.py#L296)):*
<details><summary><i>Previously, BetsyMcPhail (Betsy McPhail) wrote…</i></summary><blockquote>

Unfortunately, I think you're right. It's going to be hard to define a combination that works reliably.
</blockquote></details>

Done.

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @BetsyMcPhail and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp8zZ707lWngEEXElV0:-LpYmXrO5jmMK1uXuVYT:budk5ik) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

I will file the Drake PR that tests using this commit.
</blockquote></details>

Doing this now...

---

**Review by EricCousineau-TRI — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, 1 unresolved discussion (waiting on @BetsyMcPhail and @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp8zZ707lWngEEXElV0:-LpYoH76-9ByxlLjLkpc:be6fcf) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Doing this now...
</blockquote></details>

Filed: https://github.com/RobotLocomotion/drake/pull/12095
Waiting for CI to pass. If it's good, will merge this.

---

**Review by EricCousineau-TRI — COMMENTED** [review]
> Reviewed 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: all files reviewed, 2 unresolved discussions (waiting on @BetsyMcPhail)

___
*[tests/test_class.py, line 295 at r8](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-LpYoMQ0043otBlDjZz4:-LpYoMQ0043otBlDjZz5:b-hfq4r7) ([raw file](https://github.com/robotlocomotion/pybind11/blob/c07168a127f1a401c1fda473ffa3ca12cbe4797e/tests/test_class.py#L295)):*
> ```Python
> 
> 
> @pytest.mark.skip(reason="Generally reproducible in CPython, Python 3, non-debug, on Linux.\
> ```

nit Super minor, but for line skips, can you instead split the string literal? e.g.
```
func(arg="My text "
         "with line continuation.")
```

**Review by EricCousineau-TRI — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: 3 of 4 files reviewed, all discussions resolved (waiting on @EricCousineau-TRI)

*[a discussion](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-Lp8zZ707lWngEEXElV0:-LpZxwjA4hCOPYDb90jn:bvysoe2) (no related file):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

Filed: https://github.com/RobotLocomotion/drake/pull/12095
Waiting for CI to pass. If it's good, will merge this.
</blockquote></details>

Done. (Passed.)

---
___
*[tests/test_class.py, line 295 at r8](https://reviewable.io/reviews/robotlocomotion/pybind11/32#-LpYoMQ0043otBlDjZz4:-LpZxyyn5sA2_65DhZYe:b-896fix) ([raw file](https://github.com/robotlocomotion/pybind11/blob/c07168a127f1a401c1fda473ffa3ca12cbe4797e/tests/test_class.py#L295)):*
<details><summary><i>Previously, EricCousineau-TRI (Eric Cousineau) wrote…</i></summary><blockquote>

nit Super minor, but for line skips, can you instead split the string literal? e.g.
```
func(arg="My text "
         "with line continuation.")
```
</blockquote></details>

Done.

**Review by EricCousineau-TRI — APPROVED** [review]
> Reviewed 1 of 1 files at r9.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/pybind11/32)* status: :shipit: complete! all files reviewed, all discussions resolved

**Pull request merged**
